sopt/src/main/scala/dagr/sopt/util/Configuration.scala

```diff
@@ -0,0 +1,248 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package dagr.sopt.util
+
+import java.io.File
+import java.nio.file.{Files, Path}
+import java.text.SimpleDateFormat
+import java.time.Duration
+import java.util.Date
+
+import com.typesafe.config.ConfigException.Generic
+import com.typesafe.config.{Config, ConfigFactory, ConfigParseOptions}
+import dagr.sopt.util.Configuration.Keys
+
+import dagr.commons.util.StringUtil._
+
+import scala.collection.JavaConversions._
+import scala.collection.SortedSet
+import scala.reflect.runtime.universe.{TypeTag, typeOf}
+
+
+/**
+  * Companion object to the Configuration trait that keeps track of all configuration keys
+  * that have been requested so that they can be reported later if desired.
+  */
+object Configuration extends ConfigurationLike {
+
+  // A sorted set tracking all the configuration keys that are requested
+  private[util] val RequestedKeys = collection.mutable.TreeSet[String]()
+
+  // The global configuration instance for dagr
+  protected var _config: Config = ConfigFactory.load()
+
+  /** Implement the abstract method from the [[Configuration]] trait to return the actual config instance. */
+  override protected def config = _config
+
+  object Keys {
+    val CommandLineName = "dagr.command-line-name"
+    val ColorStatus     = "dagr.color-status"
+    val SystemPath      = "dagr.path"
+  }
+
+  /**
+    * Initialize the configuration by loading configuration from the supplied path, and combining it with
+    * configuration information from the system properties (higher priority), application.conf and
+    * reference.conf files (lower priority).
+    */
+  def initialize(path: Option[Path]): Unit = path match {
+    case None    =>
+      this._config = ConfigFactory.load()
+    case Some(p) =>
+      // setAllowMissing(false) refers to allowing the file(!) to be missing, not values within the file
+      val options = ConfigParseOptions.defaults().setAllowMissing(false)
+      val localConfig = ConfigFactory.parseFile(p.toFile, options)
+
+      // This mimics the behaviour of ConfigFactory.load() but with the localConfig sandwiched in
+      this._config = ConfigFactory.defaultOverrides()
+        .withFallback(localConfig)
+        .withFallback(ConfigFactory.defaultApplication())
+        .withFallback(ConfigFactory.defaultReference())
+        .resolve()
+  }
+
+  /** Allows initialization with a custom configuration. */
+  protected def initialize(customConfig: Config): Unit = this._config = customConfig
+
+  /** Returns a sorted set of all keys that have been requested up to this point in time. */
+  private[util] def requestedKeys: SortedSet[String] = {
```

Comment:
This should probably be public, otherwise it's not of much use.  The idea here was that dagr would track all the config keys used and be able to report the full set requested, e.g. during simulation.